### PR TITLE
core: reliably check if varlink socket has been deserialized; switch varlink server to pidref

### PR DESCRIPTION
**Pull request by YHNdnzj**
Alternative to https://github.com/systemd/systemd/pull/33817

### CODE REVIEW

File: src/core/core-varlink.c [hunk 1]
```diff
@@ -500,12 +501,17 @@ static void vl_disconnect(sd_varlink_server *s, sd_varlink *link, void *userdata
...
         int r;
 
         assert(m);
-        assert(ret);
+
+        if (m->varlink_server)
+                return 0;
+
+        if (!MANAGER_IS_SYSTEM(m))
+                return -EPROTOTYPE;
```
Comment (bluca) [inline src/core/core-varlink.c]: strange error code?

Comment (YHNdnzj) [reply src/core/core-varlink.c]: I don't think it's weird? Or `ENOEXEC` could be another candidate.

Comment (bluca) [reply src/core/core-varlink.c]: Well a user would see in the logs `Protocol wrong type for socket` which is not exactly helpful - just EINVAL would be clearer no?

Comment (YHNdnzj) [reply src/core/core-varlink.c]: Quite a bunch of stuff would fail with -EINVAL... But I changed it anyways.

**Comment by bluca** [comment]
I'll test an upgrade from the old version once the packages are available in the mkosi job

**Review by bluca — APPROVED** [review]
Ok upgrading from 247 works fine

**Pull request merged**